### DISCUSSION TITLE
Abusing the use of ProviderScope overrides is a bad or good practice?

### DISCUSSION THREAD

**Discussion by astrid-dev** [rrousselGit/riverpod#2017] [Q&A] [answered]
recently I have read a couple of examples and medium tutorial about infinite scroll list using riverpod and the basic idea is to have 3 or 4 providers and make them work together in a way all the related widgets consume them an the instances become constants, reducing the re-renders and improving performance.

the three or four providers are
- `pageDataProvider.family<PageModel, int>` this one will load every page
- `pageSizeProvider` this will consume `pageLoadProvider(0)` to get the page size (could return just a constant `int`)
- `totalCountProvider` this will consume `pageLoadProvider(0)` to get the total of elements to pass it to the list
- `elementProvider` this always throws un-implemented and should be override within the data of every element built in the `ListView.builder` builder method

the example is like this
```dart
class InifiniteList extends ConsumerWidget {
  const InifiniteList({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final totalCount = ref.watch(totalCountProvider);
    final pageSize = ref.watch(pageSizeProvider);
    return totalCount.when(
      loading: () => const LoadingWidget(),
      error: (e, s) => const ErrorWidget(),
      data: (count) => ListView.builder(
        itemCount: count, // total of elements (not just one page)
        itemBuilder: (context, index) {
          // get page size
          final pageSizeValue = pageSize.asData!.value;
          // get the page where the item is relative to the total of elements
          final page = index ~/ pageSizeValue;
          // get the element index in the page (not in the total of elements)
          final elementIndex = index % pageSizeValue;
          // individual element data
          final element = ref
              .watch(pageDataProvider(page)) // load the data of that page
              .whenData((page) => page.content.elementAt(elementIndex));

          return ProviderScope(
            overrides: [
              // override the single element provider for every item in the list
              elementProvider.overrideWithValue(element),
            ],
            // [ItemWidget] will consume [elementProvider] and render the
            // respective element, and becasue is contant it will not re render
            // when the same data is used
            child: const ItemWidget(),
          );
        },
      ),
    );
  }
}
```

Following this idea, with the `ProviderScope`, the override of individual providers and the constant widgets;
I though I could use it in every place I need to pass an individual item, instead of passing it I'll just override the provider and use a constant widget. so far it works really great, but after a couple of days of development I realised that there is no way to enforce the override of the provider (except for the un implemented exception), so if there is any other developer that wants to use that widgets they will not know the widget has that dependency and that it involves more boilerplate, this make me this that maybe is not the best practice.

```dart
// without the override we could just do
data: (item) => ItemWidget(item: item), // not a constant widget, will re render every time [data] call back is run again

// with the override and boilerplate 
data: (item) => ProviderScope(
  overrides: [
    itemProvider.overrideWithValue(item),
  ],
  child: const ItemWidget(), // const widget, will render just once even when the [data] call back is run again
),
```

**Comment by rrousselGit** [comment, marked as answer]
That's fine. A bit advanced, but fine.
